### ISSUE TITLE
Creation of dummy users when adding staff to course

### ISSUE THREAD

**Issue by jonatanschroeder** [PrairieLearn/PrairieLearn#4641]
#3020 allows the bulk addition of staff users to a course. However, as listed in https://github.com/PrairieLearn/PrairieLearn/pull/3020#issuecomment-894561115 (and following comments), the users must exist on PL (must have logged in to PL at least once in the past) for this to work. I propose we allow the creation of dummy users when the UID doesn't exist on PL, as was done before #3020, so that the user gets access granted before logging into the system. There must be some way to mitigate the potential for typos in this case.

Once this is done, it should be relatively trivial to also have a similar button to add students to a course (e.g., on the Gradebook page). See comment linked above for further discussion.

**Comment by jonatanschroeder** [comment]
Adding comments here so they're all in the same place:

> One side-effect of this PR: before this, testing in a local dev environment would create users for the instructors/TAs associated to the course based on the JSON files. This is no longer the case, so courses will have no users at all locally. There's no simple way to create users locally for testing, other than adding random students using admin queries. I don't think this is a major problem, I just want to make sure this is being considered.

> On that note, I'm considering the following use-case: I have a list of my TAs, and I want to give them all access to the course instance on PL. Previously I would just add them to the JSON file, and if they didn't exist they'd get access as soon as they logged in for the first time. Now, I can only add them to the staff list after the users have been created, which means if my list has, say, 10 TAs, and 2 fail, I'd need to ask them login, and then add them back. Would it make sense to create a provisional user (similar as before) when they're added to the course staff, maybe with a confirmation dialog to avoid issues like typos?

> > All great points. I don't see any real harm in creating provisional users (with a dialog warning to avoid typos).
> 
> The timing of this warning may be an interesting question. If done client-side, there would need to either be a blanket warning (maybe with a checkbox to add new users if they don't exist), or an ajax check if users exist. If done server-side, there would need to be a second-level page to show users that don't exist and ask if they should be added (which sounds a bit weird). The checkbox option is the simplest, the ajax looks like a pretty fancy option that could avoid typos, but may be a bit harder to implement.
> 
> > We could also do this on the Gradebook page with an "Add student" button, which would enroll a student into the course and also add them as a provisional user if they don't yet exist (again with a warning dialog). On local dev these buttons would provide a way to add students/staff for testing.
> 
> That also sounds like a good way to provide instructors the ability to control enrolment in class if they don't want to allow students to enrol themselves (alternative to LTI or other enrolment restriction methods).

**Comment by nwalters512** [comment]
See also #9217.

**Comment by jonatanschroeder** [comment]
Closing in favour of #9217.
